### ISSUE TITLE
Unable to access the SD card

### ISSUE THREAD

**Issue by DevlinDiener** [playfultechnology/audio-guestbook#52]
After compiling and uploading the code successfully my board doesn't do anything when hooked up to the phone. Naturally I plugged it in and ran the serial monitor to see if I could troubleshoot. The monitor continuously prints the message 

"Unable to access the SD card"

I have the SD card slotted directly onto the audio shield's card slot. I verified the SD card compatibility as well so I'm not sure if this is a code error or a hardware issue. Any help is appreciated!

**Comment by nrich50** [comment]
did you ever fix this issue i am having this problem aswell and nothing i have done has fixed it

**Comment by rhilyard** [comment]
I think a lot of us are having this same issue. I hope someone can tell us how to fix this. I have tried everything.

**Comment by BrockV2** [comment]
Makes 4 of us, I have also run out of ideas. Will keep trying.

**Comment by nrich50** [comment]
So after messing with this thing for over 24 hours in my case I had to change the ports since I am using the REV D audio shield and use a different SD card. Idk why my other one wouldn't work but once I put a new one in it was able to access. Bit since I have solved that problem I still have no sound coming from the speaker and no input from the mic

**Comment by Praxxxer** [comment]
@nrich50 I'm also using a REV D2 audio shield. What did you change your SDCARD_MOSI_PIN and SDCARD_SCK_PIN values to? According to https://www.pjrc.com/store/teensy3_audio.html the values should be MOSI_PIN 10 and SCK_PIN 12, but I'm not having any luck.

**Comment by BrockV2** [comment]
@Praxxxer 

Please check the PIN numbers you have quoted as they do not match the values on the teensy drawings. 

To help troubleshoot, can you please confirm if you are using Teensy 4.0 or 4.1? 

Please also check there is continuity on ground

**Comment by DevlinDiener** [comment]
Update:

TLDR: Verify the integrity of your hardware connections. Use the pins defined below if you're using Teensy 4.0 + REV D2 audio shield.

I've fixed the "unable to access SD card" issue. For reference, I'm using a Teensy 4.0 and REV D2 audio shield. Here are the pins I've assigned:

- #define SDCARD_CS_PIN    10
- #define SDCARD_MOSI_PIN  7
- #define SDCARD_SCK_PIN   14

I occasionally get this issue during testing (likely due to improper connection), but it will usually fix itself after a reboot.

My problem now is that my microphone line is not picking up any input, and my speaker doesn't produce output. I assume this is due to a bad soldering job on a few connections. I took a break from this project but am picking it back up. I'll update you all again with any useful info once I've fixed the issue.

**Comment by rhilyard** [comment]
Which fork is everyone using? I have the teensy 4.0 and Rev D2 audio shield. I used 
    #define SDCARD_CS_PIN 10
    #define SDCARD_MOSI_PIN 7
    #define SDCARD_SCK_PIN 14
but that didn
t help at all.

**Comment by nrich50** [comment]
I believe I am using the fork @DD4WH has. Have you tried a different SD card and making sure it's formatted. I think it needs to be exFAT but I'm not 100% sure. For me I changed all those pins and it still was not working until I got a new SD. And this mag be stupid but I found myself doing this. If you don't have the teensy directly mounted to the audio shield. Make sure the GND on the teensy is also connected to the audio shield GND. I am completely new to all of this so many mistakes have been made lol

**Comment by nrich50** [comment]
Still having the issue of no sound or mic input. I thought maybe I messed up the board but since many people are having it then it must be either a bad batch of audio shields or a error in programming

**Comment by rhilyard** [comment]
I have used several different types of sd cards  The gnd is connected to the audio shield. I'm completely at a lost on what to do!!

**Comment by nrich50** [comment]
@rhilyard i used this for my pins and added a extra one that is not in the original program. it may not fix your issue but its worth a shot.

#define SDCARD_CS_PIN    10
#define SDCARD_MSIO_PIN    12
#define SDCARD_MOSI_PIN  11
#define SDCARD_SCK_PIN   13

**Comment by rhilyard** [comment]
@nrich50 Thanks just gave it a try and it didn't help.  Thanks for the suggestion

**Comment by kyuubii** [comment]
I also face the same issue "Unable to access the SD card". 
I use a "Teensy 4.0" with an "Audio Shield Rev D2". 
I tried to: 
- use different SD cards (from 512 MB to 64 GB)
- use FAT, FAT32, exFAT, NTFS formats
- use different pins (10, 11, 12, 13) in the code

**Comment by markmak77** [comment]
> Update:
> 
> TLDR: Verify the integrity of your hardware connections. Use the pins defined below if you're using Teensy 4.0 + REV D2 audio shield.
> 
> I've fixed the "unable to access SD card" issue. For reference, I'm using a Teensy 4.0 and REV D2 audio shield. Here are the pins I've assigned:
> 
> * #define SDCARD_CS_PIN    10
> * #define SDCARD_MOSI_PIN  7
> * #define SDCARD_SCK_PIN   14
> 
> I occasionally get this issue during testing (likely due to improper connection), but it will usually fix itself after a reboot.
> 
> My problem now is that my microphone line is not picking up any input, and my speaker doesn't produce output. I assume this is due to a bad soldering job on a few connections. I took a break from this project but am picking it back up. I'll update you all again with any useful info once I've fixed the issue.

@DevlinDiener , Thank you for confirming that these are the correct pin associations for Teensy 4.0 + the Rev D2 audio shield. Could you also please share what you have your micro SD card formatted to? 

I have the opposite issue where I can hear the start-up "beep" if I plug headphones into the 3.5 jack on the audio shield (to test it), but I'm getting the "Unable to access the SD card" error in the serial monitor after that. Might be a bad SD card(s) - testing with some old ones I had laying around.

**Comment by markmak77** [comment]
**Update: IT WORKS!**
First of all, I'd like to confirm that the pins listed above by DevlinDiener are indeed the correct ones if you're using a Teensy 4.0 + Rev D2 audio shield - my micro SD card initialized successfully and was recognized by the controller using those exact pins. I'm using a class 10 Sandisk MicroSD card, 32 GB, formatted as FAT32 (did a full format, not quick, just in case, although that shouldn't really affect anything in theory).

Secondly, I'd like to suggest to everyone who's had this problem (or will have it in the future): **CHECK YOUR CONNECTIONS/SOLDER JOINTS**! A post [here](https://forum.pjrc.com/threads/70679-Teensy-4-0-Audio-Board-Rev-D-SD-card-problem) got me thinking this might be the issue, and it seems it was. 

I'd soldered headers onto both the Teensy (male) and the audio shield (female), _except_ for the 3 pins on the left side (GND, 0, 1) to leave them available for the button(s). Now, I know the audio shield needs a GND, but I thought it would be ok with the one on the other side, between Vin (aka 5V) and 3.3V (aka 3V), but apparently not... And, to make sure I didn't have any cold solder joints, I took my soldiering iron to each pin (on both the Teensy and the shield) once again, and it seems those 2 actions solved this issue for me. 

(cc @nrich50 , @rhilyard , @Praxxxer , @kyuubii )

PS: before doing any additional soldering, don't forget to first remove the SD card from the shield.

PPS: just noticed today that Teensy's site has SD card recommendations at the bottom of [this page](https://www.pjrc.com/store/teensy3_audio.html):

> Most SD cards are optimized for sequential access, where a camera or camcorder reads or writes a single large file. All SD cards work well for playing a single WAV file at a time.
> Cards with "A1" or "A2" rating are likely to have better performance for use with Teensy.
> Since Teensyduino 1.54, the SD library supports up to 2 TB size. Teensyduino 1.53 and earlier only supported cards up to 32GB size.
> Very old cards smaller than 8GB may have trouble with the built in SD socket on Teensy 3.5, 3.6, 4.1, but usually with fine with the SPI-based socket on the audio shield.
> The audio library includes a simple benchmark to test SD cards. Open it from File > Examples > Audio > HardwareTesting > SdCardTest.

**Comment by markabruce** [comment]
Hi everyone,

I am not yet 100% sure why, but for me the `#ifdef TEENSY_41` seemed to be the culprit...

I have a Teensy 4.0 and a Audio Shield Rev D2.

Remove the two references of `#ifdef TEENSY_41`.


**So when defining pins...**

```
#ifdef TEENSY_41
#define SDCARD_CS_PIN    BUILTIN_SDCARD
#else
#define SDCARD_CS_PIN    10
#endif
```

becomes:
```
// Define pins used by Teensy Audio Shield
#define SDCARD_CS_PIN    10
#define SDCARD_MOSI_PIN  7
#define SDCARD_SCK_PIN   14
```

**When setting MOSI and SCK pins...**

```
#ifndef TEENSY_41   // assumes that you are using the SD card slot of the AUDIO BOARD
  SPI.setMOSI(SDCARD_MOSI_PIN);
  SPI.setSCK(SDCARD_SCK_PIN);
#endif  
```

becomes:

```
  SPI.setMOSI(SDCARD_MOSI_PIN);
  SPI.setSCK(SDCARD_SCK_PIN);
```

**Comment by markmak77** [comment]
@markabruce , that's probably because you had Teensy 4.1 selected in the Arduino IDE under Tools -> Board: "Teensy x.x". To change it (i.e. define TEENSY_40 instead of TEENSY_41), select Tools -> Board: "Teensy x.x" -> Teensyduino -> Teensy 4.0.

**Comment by markabruce** [comment]
I am pretty sure I had 4.0 selected, I'll double check tomorrow.

**Comment by markabruce** [comment]
Here:

![image](https://user-images.githubusercontent.com/10707880/226093869-4a1a1430-52c8-4878-936c-92de82535e80.png)

I have Board: Teensy 4.0 and yet the `IFNDEF TEENSY_41` must somehow skip setting the pins, maybe a bug with my version of Arduino IDE (1.8.19), or the Teensy installer...

I just copied what `File > Examples > Audio > HardwareTesting > SdCardTest` was doing, which is what PRJC recommend: https://www.pjrc.com/store/teensy3_audio.html

The IFNDEF is saying **if TEENSY_41 is NOT defined, then set the SPI** - so I've no idea how the software must think TEENSY_41 is defined, therefore setting the SPI pins is not executed... is it some sort of global variable set by one of the libraries? Can't find any docs on it...

**Comment by BrockV2** [comment]
Has to be defined in the sketchbook code not the menu.

On 18 Mar 2023, at 08:36, markabruce ***@***.***> wrote:

﻿

Here:

[image]<https://user-images.githubusercontent.com/10707880/226093869-4a1a1430-52c8-4878-936c-92de82535e80.png>

I have Board: Teensy 4.0 and yet the IFNDEF TEENSY_41 must somehow skip settings the pins, maybe a bug with my version of Arduino IDE (1.8.19), or the Teensy installer...

I just copied what File > Examples > Audio > HardwareTesting > SdCardTest was doing, which is what PRJC recommend: https://www.pjrc.com/store/teensy3_audio.html

The IFNDEF is saying if TEENSY_41 is NOT defined, then set the SPI - so I've not idea how the software must think TEENSY_41 is defined, is it some sort of global variable set by one of the libraries? Can't find any docs on it...

**Comment by markabruce** [comment]
> Has to be defined in the sketchbook code not the menu. 

Aha! Apologies, I am new to this... Thank you.

I was being a fool and I hadn't removed #define TEENSY_41, I thought I had searched for every reference to it, but obviously didn't! The forked code even mentions to do this in the [README.md](https://github.com/DD4WH/audio-guestbook/blob/main/README.md).

I suspect though, that this could be the issue a lot of other people are facing. Search for every line of `TEENSY_41` and comment it out or remove it, it is not required if you're on a TEENSY 4.0.
